include/verilated_vpi.cpp

```diff
@@ -2299,13 +2303,29 @@ bool vl_check_format(const VerilatedVar* varp, const p_vpi_value valuep, const c
     return status;
 }
 
+void strprintf(std::string& buffer, char const* fmt, ...) {
+    va_list args, args_copy;
+    va_start(args, fmt);
+    buffer.clear();
+    // make copy of args since we may need to call vsnprintf more than once
+    va_copy(args_copy, args);
+    // try vspnrintf in existing buffer
+    int result = vsnprintf(const_cast<char*>(buffer.data()), buffer.capacity(), fmt, args_copy);
+    va_end(args_copy);
+    // ensure(result >= 0)
+    int const required = result + 1;  // returned size doesn't include NUL terminator
+    // if there wasn't enough space, reallocate and try again
+    if (buffer.capacity() < required) {
+        buffer.reserve(required);
```

Comment:
```suggestion
        buffer.reserve(required * 2);
```

Can greatly reduce reallocations.